Initial work on adding possible rotation of infinity sign.

Filco306 opened the pull request:
Hello! 

Thank you for a great repository! I wanted and needed to rotate the infinity sign, so I though I might add it here. 

Basically, I add an argument to allow for rotating an infinity sign. As an example, using `angle=np.pi/2` will generate an eight. 

Best,
Filip

Filco306 commented:
I think the conflicts arose after the recent pull request was merged in.

Filco306 commented:
Fixed the merge conflicts; both the `seed` PR and my PR are integrated now.

sauln commented on tadasets/shapes.py:
```diff
@@ -180,5 +181,11 @@ def infty_sign(n=100, noise=None, seed=None):
+        try:
+            assert angle >= -np.pi and angle <= 2*np.pi
+        except AssertionError:
+            raise RotationAngleNotInRangeError(angle, 0, "pi")
```
I think the custom error hear is little overkill. Would we lose anything by having a single line?

```
assert angle >= -np.pi and angle <= 2*np.pi, "Angle {angle} not in range. Angle should be in the range {min_angle} <= angle <= {max_angle}".format(...)
```

Filco306 replied:
No, great point! I will fix that.

sauln commented on tadasets/rotate.py:
```diff
@@ -0,0 +1,35 @@
+    except AssertionError:
+        raise ValueError("Error: data has {} dimensions, but should only be 2. ".format(d.shape[1]))
+
+    if centered_around_zero is False:
```
Since this is all new, could you remove the `cenetered_around_zero` stuff?  I'd rather leave this out and just add it once implemented rather than provide an interface without implementation.

Filco306 replied:
Yeah, sure! Good point! should I remove the `ValueError` as well?

sauln replied:
I think so.  Though, if you have a good case for keeping it, I'm all ears.

Filco306 replied:
Well, my thoughts around `ValueError` revolved around using this rotation function for other functions as well. I think that is simply my main point; it would ensure the rotation is not incorrectly used for other dimensions than 2. Maybe we should keep it? Although `centered_around_zero`  is probably removeable.

Filco306 replied:
Now I removed `centered_around_zero`, but I kept the `ValueError` :)

sauln reviewed (changes requested):
Overall, I think it looks good and this will be a nice addition to the library. Thank you for the contribution! 

There are a couple of things I commented on that I think could be cleaned up before incorporating it. 

Also, could you add a test or two for the rotate function and the infinity sign function?

sauln commented on test/test_shapes.py:
```diff
@@ -93,8 +93,24 @@ def test_n(self):
+        assert t.shape[0] == 345
+
+    def test_rotation(self):
+        try:
```
For the testing here, checkout pytest's `pytest.raises` method documented here https://docs.pytest.org/en/stable/assert.html

Filco306 replied:
Done!

sauln commented:
Looks good! just a small tweak to leverage `pytest.raises` in the tests and we're good to go 👍

Filco306 commented:
This is actually great, I didn't know about `pytest.raises`. Great feedback on the code, thank you for that! :D

sauln commented on test/test_shapes.py:
```diff
@@ -104,13 +104,8 @@ def test_n(self):
 
     def test_rotation(self):
-        try:
+        with pytest.raises(AssertionError) as ae:
```
This will only catch the first one.  I don't think the second call with angle=300 will ever hit the `pytest.raises` if the first one throws.  Could you wrap both in their own `pytest.raises` block?

Filco306 replied:
Yes! Absolutely!

Filco306 replied:
Done!

sauln commented on tadasets/swiss_holes_helpers.py:
```diff
@@ -0,0 +1,54 @@
+    radiuses: np.ndarray with shape (n_holes, )
+        All the radiuses of all the holes.
+    """
+    return any(np.apply_along_axis(np.linalg.norm,1,row - centers) <= radiuses) is False
```
not sure if it's the case, but would `np.any` be faster than the native python `any`?

bdice replied:
They also behave differently.
```python
import numpy as np
data = [[False]]
np.any(data)  # False
any(data)  # True
```
I recommend `return not np.any(...)` instead of `return any(...) is False`.

sauln commented on tadasets/shapes.py:
```diff
@@ -185,3 +186,37 @@ def infty_sign(n=100, noise=None, angle=None, seed=None):
+    n_holes: int
+        number of holes in return data set.
+    d: int
+        number of dimensions
```
Could you add `non_overlapping` to the docstring parameter list?

sauln commented on tadasets/swiss_holes_helpers.py:
```diff
@@ -0,0 +1,54 @@
+
+    centers: np.ndarray with shape (n_holes, d)
+        All the center coordinates of a ball.
+    radiuses: np.ndarray with shape (n_holes, )
```
spelling radii?

Filco306 replied:
Haha yes, sorry, english is my second language :P

Filco306 replied:
Changed this now!

sauln reviewed (commented):
It looks like both PRs are merged into 1 now?  I'm okay with 1 big one if that was your intention. 

A couple of small tasks before we merge it:
* Could you run `black` over the files you changed to get the formatting in line? 
* Could you test the docs to make sure they show the new changes you've made to the api?
* Could you bump the version in `tadasets/_version.py`?
Thank you!

Filco306 commented:
> It looks like both PRs are merged into 1 now? I'm okay with 1 big one if that was your intention.
> 
> A couple of small tasks before we merge it:
> 
> * Could you run `black` over the files you changed to get the formatting in line?
> * Could you test the docs to make sure they show the new changes you've made to the api?
> * Could you bump the version in `tadasets/_version.py`?
>   Thank you!

Hello! I accidentally merged the other PR into this, but I reverted it, so this only contains the changes with the infty sign now :) 

- Black has been run! I use `pre-commit` with black and flake8. If you want to, I can add those configurations to this repo if you want to? They help ALOT in making code cleaner :) 
- I tested the docs. I am new to sphinx but I added it and ran `make html` inside the `docs` folder. It gets an error: 

```
make html
Running Sphinx v3.3.1

Configuration error:
There is a programmable error in your configuration file:

Traceback (most recent call last):
  File "/usr/local/lib/python3.8/site-packages/sphinx/config.py", line 319, in eval_config_file
    execfile_(filename, namespace)
  File "/usr/local/lib/python3.8/site-packages/sphinx/util/pycompat.py", line 89, in execfile_
    exec(code, _globals)
  File "/Users/filipcornell/Desktop/code/forks/tadasets/docs/conf.py", line 7, in <module>
    from sktda_docs_config import *
ModuleNotFoundError: No module named 'sktda_docs_config'

make: *** [html] Error 2
```
Again, I am new to `sphinx` so it might be a noob error.
- By bumping, do you mean changing it to `0.0.5`?

Again, your feedback is great and thank you for your guidance! Keep pushing for these small details, they are great for me to learn new things!

sauln commented:
> ModuleNotFoundError: No module named 'sktda_docs_config'

Try installing that library and then it should be good to go. 

> If you want to, I can add those configurations to this rep
That would be great 👍 

> By bumping, do you mean changing it to 0.0.5?
Let's do 0.1.0!

Filco306 commented:
I bumped it now, but the changes does not appear in the documentation. `angle` and `seed` do not appear. Hmm

sauln commented:
hmm indeed.. When you ran `make html`, did it generate a `_build` dir with `index.html` in it?

Filco306 commented:
> hmm indeed.. When you ran `make html`, did it generate a `_build` dir with `index.html` in it?

It did. But it does not generate it with the new parameters :/ 

For `infty_sign`, I get this: 

```HTML
<span class="n">__all__</span> <span class="o">=</span> <span class="p">[</span><span class="s2">&quot;torus&quot;</span><span class="p">,</span> <span class="s2">&quot;dsphere&quot;</span><span class="p">,</span> <span class="s2">&quot;sphere&quot;</span><span class="p">,</span> <span class="s2">&quot;swiss_roll&quot;</span><span class="p">,</span> <span class="s2">&quot;infty_sign&quot;</span><span class="p">]</span>


<span class="c1">## TODO: Make a base class that controls `ambient` and `noise`.</span>
<span class="k">class</span> <span class="nc">Shape</span><span class="p">:</span>
    <span class="k">def</span> <span class="fm">__init__</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="k">pass</span>


<div class="viewcode-block" id="dsphere"><a class="viewcode-back" href="../../reference/stubs/tadasets.dsphere.html#tadasets.dsphere">[docs]</a><span class="k">def</span> <span class="nf">dsphere</span><span class="p">(</span><span class="n">n</span><span class="o">=</span><span class="mi">100</span><span class="p">,</span> <span class="n">d</span><span class="o">=</span><span class="mi">2</span><span class="p">,</span> <span class="n">r</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span> <span class="n">noise</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">ambient</span><span class="o">=</span><span class="kc">None</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Sample `n` data points on a d-sphere.</span>

<span class="sd">    Parameters</span>
<span class="sd">    -----------</span>
<span class="sd">    n : int</span>
<span class="sd">        Number of data points in shape.</span>
<span class="sd">    r : float</span>
<span class="sd">        Radius of sphere.</span>
<span class="sd">    ambient : int, default=None</span>
<span class="sd">        Embed the sphere into a space with ambient dimension equal to `ambient`. The sphere is randomly rotated in this high dimensional space.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">data</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">random</span><span class="o">.</span><span class="n">randn</span><span class="p">(</span><span class="n">n</span><span class="p">,</span> <span class="n">d</span><span class="o">+</span><span class="mi">1</span><span class="p">)</span>

    <span class="c1"># Normalize points to the sphere</span>
    <span class="n">data</span> <span class="o">=</span> <span class="n">r</span> <span class="o">*</span> <span class="n">data</span> <span class="o">/</span> <span class="n">np</span><span class="o">.</span><span class="n">sqrt</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">sum</span><span class="p">(</span><span class="n">data</span><span class="o">**</span><span class="mi">2</span><span class="p">,</span> <span class="mi">1</span><span class="p">)[:,</span> <span class="kc">None</span><span class="p">])</span> 

    <span class="k">if</span> <span class="n">noise</span><span class="p">:</span> 
        <span class="n">data</span> <span class="o">+=</span> <span class="n">noise</span> <span class="o">*</span> <span class="n">np</span><span class="o">.</span><span class="n">random</span><span class="o">.</span><span class="n">randn</span><span class="p">(</span><span class="o">*</span><span class="n">data</span><span class="o">.</span><span class="n">shape</span><span class="p">)</span>

    <span class="k">if</span> <span class="n">ambient</span><span class="p">:</span>
        <span class="k">assert</span> <span class="n">ambient</span> <span class="o">&gt;</span> <span class="n">d</span><span class="p">,</span> <span class="s2">&quot;Must embed in higher dimensions&quot;</span>
        <span class="n">data</span> <span class="o">=</span> <span class="n">embed</span><span class="p">(</span><span class="n">data</span><span class="p">,</span> <span class="n">ambient</span><span class="p">)</span>



    <span class="k">return</span> <span class="n">data</span></div>

```

Might have missed some html, but this captures the main problem

sauln commented:
Okay, no worries. Everything else looks good to go.  I'll try taking a look between merging this and getting the second holes in swiss PR merged.

Is this all ready to go for you?

Filco306 commented:
Hello! 

Please excuse my late reply. I added my pre-commit linting now that uses `.flake8` and `black`. I also fixed some minor issues that I found in different places, such as variables that were assigned but never used (those assignments were commented out) and assured that it would work. It passes all the tests and is now better code :) 

On the infty-rotation, it is all done and ready to be merged :) 

Thank you for all your trouble with this! I learnt a lot by contributing here! 

Cheers,
Filip

sauln commented:
As with the other one, what are your thoughts on merging it?  It's been a bit so I can't remember where it was left.  Ready to go?

Filco306 commented:
Hello @sauln , please excuse my very late reply! This one is ready to go. I think the other one, although it is interesting, it does need some more work because it cannot scale with as the number of dimensions scale. 

Again, sorry for such a late reply, but this one is ready to go :)

sauln commented:
Sweet! Here it goes!
